### PR TITLE
Support `axis=None` in sparse min/max

### LINKED ISSUE #3506: Sparse min/max does not support `axis=None`
Rel #3474.

Sparse min/max does not support `axis=None`, which is the default value for the parameter.

```Python
>>> import cupy
>>> import cupyx
>>> m = cupy.random.rand(100).reshape(10, 10)
>>> m[m < 0.95] = 0
>>> m = cupyx.scipy.sparse.csr_matrix(m)
>>> m.min()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ext-mtakagi/cupy/cupyx/scipy/sparse/data.py", line 284, in min
    return self._min_or_max(axis, out, cupy.min, sum_duplicates, nonzero)
  File "/home/ext-mtakagi/cupy/cupyx/scipy/sparse/data.py", line 163, in _min_or_max
    axis, min_or_max, sum_duplicates, non_zero)
  File "/home/ext-mtakagi/cupy/cupyx/scipy/sparse/data.py", line 139, in _min_or_max_axis
    N = self.shape[axis]
TypeError: tuple indices must be integers or slices, not NoneType
```

### PR DESCRIPTION
**Pull request by takagi**
Close #3506.

This PR fixes sparse min/max to support `axis=None`.

### CODE REVIEW

File: cupyx/scipy/sparse/data.py [hunk 1]
```diff
@@ -156,6 +158,20 @@ def _min_or_max(self, axis, out, min_or_max, sum_duplicates, non_zero):
...
+            if 0 in self.shape:
+                raise ValueError("zero-size array to reduction operation")
+
+            zero = self.dtype.type(0)
+            if self.nnz == 0:
+                return zero
+            if sum_duplicates:
+                self.sum_duplicates()
+            m = min_or_max(self.data.ravel())
+            if self.nnz != numpy.prod(self.shape):
```
Comment (toslunar) [inline cupyx/scipy/sparse/data.py:171]: How about `cupy.core.internal.prod`?

File: cupyx/scipy/sparse/data.py [hunk 2]
```diff
@@ -156,6 +158,20 @@ def _min_or_max(self, axis, out, min_or_max, sum_duplicates, non_zero):
...
+                m = min_or_max(zero, m)
```
Comment (toslunar) [inline cupyx/scipy/sparse/data.py:172]: Could you investigate the `nonzero` option?

Comment (leofang) [reply cupyx/scipy/sparse/data.py:172]: `min_or_max(zero, m)` seems odd -- under the hood it's `cupy.min(zero, m)`, which doesn't make sense...

File: cupyx/scipy/sparse/data.py [hunk 3]
```diff
@@ -156,6 +157,27 @@ def _min_or_max(self, axis, out, min_or_max, sum_duplicates, non_zero):
...
+            m = min_or_max(self.data.ravel())
+            if non_zero:
+                return m
+            if self.nnz != internal.prod(self.shape):
+                if min_or_max is cupy.min:
+                    m = cupy.minimum(zero, m)
+                elif min_or_max is cupy.max:
+                    m = cupy.maximum(zero, m)
+                else:
+                    assert False
```
Comment (leofang) [inline cupyx/scipy/sparse/data.py:178]: Maybe `raise RuntimeError`?

File: cupyx/scipy/sparse/data.py [hunk 4]
```diff
@@ -156,6 +157,27 @@ def _min_or_max(self, axis, out, min_or_max, sum_duplicates, non_zero):
 
         util.validateaxis(axis)
 
+        if axis is None:
+            if 0 in self.shape:
+                raise ValueError("zero-size array to reduction operation")
+
+            zero = self.dtype.type(0)
```
[the remark this answers is not in the document]
Comment (toslunar) [reply cupyx/scipy/sparse/data.py:164]: I agree.  To be more precise:
- CuPy does not convert a 0d result into a scalar;
- `return zero` on L166 violates the policy, while `m = cupy.minimum(zero, m)` is OK.

File: cupyx/scipy/sparse/data.py (hunk 2 above)
Comment (takagi) [reply cupyx/scipy/sparse/data.py]: I fixed to use `cupy.minimum` and `cupy.maximum`.

File: cupyx/scipy/sparse/data.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (takagi) [reply cupyx/scipy/sparse/data.py]: I've fixed it!

File: cupyx/scipy/sparse/data.py [hunk 5]
```diff
@@ -156,6 +157,27 @@ def _min_or_max(self, axis, out, min_or_max, sum_duplicates, non_zero):
...
+            if self.nnz == 0:
+                return zero
+            if sum_duplicates:
+                self.sum_duplicates()
+            m = min_or_max(self.data.ravel())
```
[the remark this answers is not in the document]
Comment (takagi) [reply cupyx/scipy/sparse/data.py]: Removed!

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py [hunk 6]
```diff
@@ -903,7 +903,34 @@ def test_getnnz(self, xp, sp):
...
+
+    def test_min_sparse_axis_none(self):
+        dm_data = numpy.random.random((10, 20))
+        dm_data[dm_data < 0.95] = 0
+
+        dm_data = scipy.sparse.csc_matrix(dm_data)
+        cp_matrix = sparse.csc_matrix((cupy.array(dm_data.data),
+                                       cupy.array(dm_data.indices),
+                                       cupy.array(dm_data.indptr)),
+                                      shape=(10, 20))
```
Comment (leofang) [inline tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py:916]: I wonder if we could just call `cp_matrix = sparse.csc_matrix(dm_data)`? (I understand this matches the style in the other tests, just wondering).

Comment (takagi) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py:916]: Yes, I think so.

**Comment by leofang** [comment]
Just to expedite the process...
Jenkins, test this please

**Pull request merged**